va/caa.go

```diff
@@ -116,7 +117,61 @@ func parseResults(results []caaResult) (*CAASet, error) {
 	return nil, nil
 }
 
-func (va *ValidationAuthorityImpl) parallelCAALookup(ctx context.Context, name string, lookuper func(context.Context, string) ([]*dns.CAA, error)) []caaResult {
+// If the fqdn has a direct alias among the list of CNAMEs, return it; otherwise
+// return "". Note that RFC 6844 only asks for tree-climbing on direct aliases,
+// not recursive aliases.
+func findAlias(source string, cnames []*dns.CNAME) string {
+	for _, cname := range cnames {
+		if cname.Header().Name == source {
+			return cname.Target
+		}
+	}
+	return ""
+}
+
+func parent(fqdn string) string {
+	labels := strings.Split(fqdn, ".")
+	if len(labels) > 1 {
+		return strings.Join(labels[1:], ".")
+	} else {
+		return ""
+	}
+}
+
+// Implement pre-erratum 5065 style tree-climbing CAA. Note: a strict
+// interpretation of pre-5065 indicates a linear lookup path - if there is any
+// CNAME at all, that precludes further tree-climbing on the original FQDN. This
+// is clearly wrong. We implement a hybrid approach that is strictly more
+// conservative: We always do full tree-climbing on the original FQDN (by virtue
+// of parallelCAALookup. When the LegacyCAA flag is enabled, we also
+// do linear tree climbing on single-level aliases.
+func (va *ValidationAuthorityImpl) treeClimbingLookupCAA(ctx context.Context, fqdn string) ([]*dns.CAA, []*dns.CNAME, error) {
+	target := fqdn
+	// Limit CNAME chasing to break CNAME loops
+	for i := 0; i < 8 && target != ""; i++ {
+		caas, cnames, err := va.dnsClient.LookupCAA(ctx, target)
+		if err != nil {
+			return nil, nil, err
+		} else if len(caas) > 0 {
+			return caas, nil, nil
+		} else if len(cnames) > 0 {
+			target = findAlias(target, cnames)
+		} else if i == 0 {
```

Comment:
Minor nit: It might improve readability slightly if the `i == 0` case with a `return` was handled above the `else if len(cnames) > 0` clause above. That way the two clauses without a `return` are immediately adjacent.

Reply:
This was obsoleted by later changes.